src/parser/wast-parser.cpp

```diff
@@ -0,0 +1,552 @@
+    auto i = in.takeI32();
+    if (!i) {
+      return in.err("expected i32");
+    }
+    if (!in.takeRParen()) {
+      return in.err("expected end of i32.const");
+    }
+    return Literal(*i);
+  }
+
+  if (in.takeSExprStart("i64.const"sv)) {
+    auto i = in.takeI64();
+    if (!i) {
+      return in.err("expected i64");
+    }
+    if (!in.takeRParen()) {
+      return in.err("expected end of i64.const");
+    }
+    return Literal(*i);
+  }
+
+  if (in.takeSExprStart("f32.const"sv)) {
+    auto f = in.takeF32();
+    if (!f) {
+      return in.err("expected f32");
+    }
+    if (!in.takeRParen()) {
+      return in.err("expected end of f32.const");
+    }
+    return Literal(*f);
+  }
+
+  if (in.takeSExprStart("f64.const"sv)) {
+    auto f = in.takeF64();
+    if (!f) {
+      return in.err("expected f64");
+    }
+    if (!in.takeRParen()) {
+      return in.err("expected end of f64.const");
+    }
+    return Literal(*f);
+  }
+
+  if (in.takeSExprStart("v128.const"sv)) {
+    Literal vec;
+    if (in.takeKeyword("i8x16"sv)) {
+      std::array<Literal, 16> lanes;
+      for (int i = 0; i < 16; ++i) {
+        auto n = in.takeI8();
+        if (!n) {
+          return in.err("expected i8");
+        }
+        lanes[i] = Literal(uint32_t(*n));
+      }
+      vec = Literal(lanes);
+    } else if (in.takeKeyword("i16x8"sv)) {
+      std::array<Literal, 8> lanes;
+      for (int i = 0; i < 8; ++i) {
+        auto n = in.takeI16();
+        if (!n) {
+          return in.err("expected i16");
+        }
+        lanes[i] = Literal(uint32_t(*n));
+      }
+      vec = Literal(lanes);
+    } else if (in.takeKeyword("i32x4"sv)) {
+      std::array<Literal, 4> lanes;
+      for (int i = 0; i < 4; ++i) {
+        auto n = in.takeI32();
+        if (!n) {
+          return in.err("expected i32");
+        }
+        lanes[i] = Literal(*n);
+      }
+      vec = Literal(lanes);
+    } else if (in.takeKeyword("i64x2"sv)) {
+      std::array<Literal, 2> lanes;
+      for (int i = 0; i < 2; ++i) {
+        auto n = in.takeI64();
+        if (!n) {
+          return in.err("expected i32");
+        }
+        lanes[i] = Literal(*n);
+      }
+      vec = Literal(lanes);
+    } else if (in.takeKeyword("f32x4"sv)) {
+      std::array<Literal, 4> lanes;
+      for (int i = 0; i < 4; ++i) {
+        auto f = in.takeF32();
+        if (!f) {
+          return in.err("expected f32");
+        }
+        lanes[i] = Literal(*f);
+      }
+      vec = Literal(lanes);
+    } else if (in.takeKeyword("f64x2"sv)) {
+      std::array<Literal, 2> lanes;
+      for (int i = 0; i < 2; ++i) {
+        auto f = in.takeF64();
+        if (!f) {
+          return in.err("expected f64");
+        }
+        lanes[i] = Literal(*f);
+      }
+      vec = Literal(lanes);
+    } else {
+      return in.err("unexpected vector shape");
+    }
+    if (!in.takeRParen()) {
+      return in.err("expected end of v128.const");
+    }
+    return vec;
+  }
```

Comment:
There seems to be duplication here of the core parser logic for these constants etc. - can that be shared somehow?

Reply:
I initially implemented this by calling `parseExpression`, then translating the parsed constant `Expression*` into a literal. I forget why reimplementing the parsing ended up being easier, so I'll try that again.

Reply:
Ok, this reuses the existing parser now. PTAL.